elastic-agent-client.proto

```diff
@@ -9,86 +9,87 @@ package proto;
-// `CONFIGURING`.
-message StateObserved {
-    // Status codes for the current state.
-    enum Status {
-        // Application is starting.
-        STARTING = 0;
-        // Application is currently configuring.
-        CONFIGURING = 1;
-        // Application is in healthy state.
-        HEALTHY = 2;
-        // Application is working but in a degraded state.
-        DEGRADED = 3;
-        // Application is failing completely.
-        FAILED = 4;
-        // Application is stopping.
-        STOPPING = 5;
-    }
-    // Token that is used to unique identify the application to agent. When agent started this
-    // application it would have provided it this token.
-    string token = 1;
+// doesn't match the expected `config_state_idx` that Elastic Agent expects, the unit is always marked as
+// `CONFIGURING` and a new `UnitExpected` will be sent to so it can have the latest configuration.
+message UnitObserved {
+    // Unique ID of the unit.
+    string id = 1;
+    // Unit type.
+    UnitType type = 2;
     // Current index of the applied configuration.
     uint64 config_state_idx = 2;
-    // Status code.
-    Status status = 3;
+    // State of unit.
+    State status = 3;
     // Message for the health status.
     string message = 4;
     // JSON encoded payload for the status.
     string payload = 5;
 }
 
-// A state expected message is streamed from the Elastic Agent to the application informing the application
-// what Elastic Agent expects the applications state to be.
-message StateExpected {
-    enum State {
-        // Expects that the application is running.
-        RUNNING = 0;
-        // Expects that the application is stopping.
-        STOPPING = 1;
-    }
-    // Expected state of the application.
-    State state = 1;
-    // Index of the either current configuration or new configuration provided.
-    uint64 config_state_idx = 2;
-    // Resulting configuration. (If the application already has the current `config_state_idx` this
-    // will be empty.)
-    string config = 3;
+// Observed statuses and configuration for defined units.
+//
+// In the case that a unit is missing from the observation then the Elastic Agent will mark that missing unit
+// as `STARTING` and send a new `UnitExpected` for the missing input.
+message CheckinObserved {
+    // Token that is used to unique identify the connection to the Elastic Agent.
+    string token = 1;
+    // Units observed state.
+    repeated UnitObserved units = 2;
 }
 
-// A action request is streamed from the Elastic Agent to the application so an action can be performed
-// by the connected application.
+// A action request is streamed from the Elastic Agent to a unit so an action can be performed by that unit.
 message ActionRequest {
     // Unique ID of the action.
     string id = 1;
+    // Unique ID of the unit.
+    string unit_id = 2;
     // Name of the action.
-    string name = 2;
+    string name = 3;
```

Comment:
The backwards compatibility rules are summarized here: https://developers.google.com/protocol-buffers/docs/proto3#updating

Adding a new enum entry is a backwards compatible change.

Reply:
I disagree, an input can easily create any action they would like. By making this an enum that means the protocol needs to be updated every time 1 single input adds an action. We don't want that type of require on the protocol.

Reply:
I agree with @blakerouse, concerning the protocol. 

I have two proposal, to bring more strictness to the code:

- We could allow developers to define in the spec file (next to the binary) what actions they are supporting
- When we do the handshake we could receive a list of supported actions from the running process. This would not require changes on the protocol but would add more `strictness` here.

Reply:
I think defining them in the spec file is a possibility but I don't know why it is really needed. Elastic Agent will only send an action that is targeted from Fleet to a specific input, for that action to occur in needs to be implemented in Fleet (or maybe as a package action in the future).

If we send an action to a process that doesn't understand that action it can just respond with an error, which is how the library that is in this repository already works. I think adding strictness even at the spec file level just seems like overhead that then the engineers will have to remember to add. To me the open question is why? I don't see a compelling reason.